naming/src/main/java/com/alibaba/nacos/naming/web/ClientAttributesFilter.java

```diff
@@ -0,0 +1,126 @@
+import com.alibaba.nacos.naming.core.v2.client.manager.ClientManager;
+import com.alibaba.nacos.naming.misc.UtilsAndCommons;
+import org.springframework.beans.factory.annotation.Autowired;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+/**
+ * <p>
+ * collect client attributes for 1.x.
+ * </p>
+ *
+ * @author hujun
+ */
+public class ClientAttributesFilter implements Filter {
+    
+    private static final String BEAT_URI = "/beat";
+    
+    private static final String IP = "ip";
+    
+    private static final String PORT = "port";
+    
+    private static final String ZERO = "0";
+    
+    @Autowired
+    private ClientManager clientManager;
+    
+    public static ThreadLocal<ClientAttributes> threadLocalClientAttributes = new ThreadLocal<>();
+    
+    @Override
+    public void doFilter(ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+            throws IOException {
+        HttpServletRequest request = (HttpServletRequest) servletRequest;
+        HttpServletResponse resp = (HttpServletResponse) servletResponse;
+        try {
+            if ((UtilsAndCommons.NACOS_SERVER_CONTEXT + UtilsAndCommons.NACOS_NAMING_CONTEXT
+                    + UtilsAndCommons.NACOS_NAMING_INSTANCE_CONTEXT).equals(request.getRequestURI())
+                    && request.getMethod().equals(HttpMethod.POST)) {
+                //register
+                ClientAttributes requestClientAttributes = getClientAttributes(request);
+                threadLocalClientAttributes.set(requestClientAttributes);
+            } else if ((UtilsAndCommons.NACOS_SERVER_CONTEXT + UtilsAndCommons.NACOS_NAMING_CONTEXT
+                    + UtilsAndCommons.NACOS_NAMING_INSTANCE_CONTEXT + BEAT_URI).equals(request.getRequestURI())) {
+                //beat
+                String ip = WebUtils.optional(request, IP, StringUtils.EMPTY);
+                int port = Integer.parseInt(WebUtils.optional(request, PORT, ZERO));
+                String clientId = IpPortBasedClient.getClientId(ip + InternetAddressUtil.IP_PORT_SPLITER + port, true);
+                IpPortBasedClient client = (IpPortBasedClient) clientManager.getClient(clientId);
+                if (client != null) {
+                    ClientAttributes requestClientAttributes = getClientAttributes(request);
+                    //update clientAttributes,when client version attributes is null,then update.
+                    if (canUpdateClientAttributes(client, requestClientAttributes)) {
+                        client.setAttributes(requestClientAttributes);
+                    }
+                }
+            }
+            filterChain.doFilter(request, servletResponse);
+            if (threadLocalClientAttributes.get() != null) {
```

Comment:
remove thread local move to final. otherwise will casue memory leak when many exception thrown.

Reply:
ok,i will fix it